flybywiresim/aircraft
# fix(mcdu): improve PERF pages

Fixes #8190 

## Summary of Changes

- Various layout improvements of the PERF CLB, PERF CRZ and PERF DES pages.
- Implement option to preselect a managed descent speed.
- Implement CHECK SPEED MODE message.
- Only allow preselecting a selected CLB speed in CAS.
- Allow preselecting a cruise speed as CAS or Mach.
- Show cruise speed as CAS below FL250 and as Mach above FL250.
- Implement option to modify PRED TO altitude on PERF CLB and PERF DES pages.
- Add DES predictions in selected modes.

## Screenshots (if necessary)


![image](https://github.com/flybywiresim/aircraft/assets/22713769/d313a85b-5b89-4009-bc03-62a7130f2239)
![image](https://github.com/flybywiresim/aircraft/assets/22713769/54694953-6a47-448c-8c04-8da340305b1e)
![image](https://github.com/flybywiresim/aircraft/assets/22713769/6083304c-a9f7-4a7e-b1d8-72f630e813c2)

## References





## Additional context

Limitations:
- No EXPEDITE predictions yet on the PERF DES page.


Discord username (if different from GitHub): BlueberryKing

## Testing instructions

- CHECK SPEED MODE
  - This message should appear when you the Climb or Cruise phase starts with a selected speed on the FCU, but no preselected speed on the respective CLB or CRZ page.
- CLB speed preselection
  - Check that you can preselect a speed between 100 kts and 350 kts. Check that you cannot preselect a Mach number.
  - Check that you can clear the preselected speed with the CLR key. Check that you can also clear it by clicking the LSK next to the managed speed target
  - Check that the preselected speed activates when the aircraft enters the CLB phase and that you do not see a CHECK SPEED MODE message
- CRZ speed preselection
  - Check that you can preselect a speed between 100 kts and 350 kts or a Mach number.
  - Check that you can clear the preselected speed with the CLR key. Check that you can also clear it by clicking the LSK next to the managed speed target
  - Check that the preselected speed activates when the aircraft enters the CRZ phase and that you do not see a CHECK SPEED MODE message
- DES speed preselection
  - Check that you can enter a Mach number as "MM/" or a CAS as "/CAS" on the DES page. This should be the managed target speed when the DES phase activates
  - Check that you can enter a CAS as "CAS", the Mach number should automatically be filled in as 0.80.
  - Check that you can clear this field to return to the precomputed Mach/CAS based on the cost index.
- CLB and DES predictions on the PERF page
  - In the CLB or DES phases respectively, check that you can modify the altitudes in the PRED TO field. On the CLB phase, any altitude between the aircraft's current altitude and the cruise altitude should be allowed. On the DES page, any altitude below the aircraft's current altitude should be allowed. Using CLR on the PRED TO field should return to the FCU altitude.
-  Speed shown on the CRZ page
   - The speed shown on the CRZ page should be a Mach number if the cruise level is above FL250, a CAS otherwise.


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
